### ISSUE TITLE
[oximeter-agent] "ERRO slog-async: logger dropped messages due to channel overflow"

### ISSUE THREAD

**Issue by smklein** [oxidecomputer/omicron#971]
I've seen the block of messages a fair bit in both testing and development. Easy repro: run `cargo test -p omicron-nexus integration_tests`.

```
Apr 25 02:00:39.762 ERRO slog-async: logger dropped messages due to channel overflow, count: 171, producer_id: a6458b7d-87c3-4483-be96-854d814c20de, component: collection-task, collector_id: 39e6175b-4df2-4730-b11d-cbc1e60a2e78, component: oximeter-agent
Apr 25 02:00:39.762 ERRO slog-async: logger dropped messages due to channel overflow, count: 1, producer_id: a6458b7d-87c3-4483-be96-854d814c20de, component: collection-task, collector_id: 39e6175b-4df2-4730-b11d-cbc1e60a2e78, component: oximeter-agent
Apr 25 02:00:39.773 ERRO slog-async: logger dropped messages due to channel overflow, count: 17, producer_id: a6458b7d-87c3-4483-be96-854d814c20de, component: collection-task, collector_id: 39e6175b-4df2-4730-b11d-cbc1e60a2e78, component: oximeter-agent
Apr 25 02:00:39.774 ERRO slog-async: logger dropped messages due to channel overflow, count: 1, producer_id: a6458b7d-87c3-4483-be96-854d814c20de, component: collection-task, collector_id: 39e6175b-4df2-4730-b11d-cbc1e60a2e78, component: oximeter-agent
Apr 25 02:00:39.783 ERRO slog-async: logger dropped messages due to channel overflow, count: 141, producer_id: a6458b7d-87c3-4483-be96-854d814c20de, component: collection-task, collector_id: 39e6175b-4df2-4730-b11d-cbc1e60a2e78, component: oximeter-agent
Apr 25 02:00:39.786 ERRO slog-async: logger dropped messages due to channel overflow, count: 5, producer_id: a6458b7d-87c3-4483-be96-854d814c20de, component: collection-task, collector_id: 39e6175b-4df2-4730-b11d-cbc1e60a2e78, component: oximeter-agent
Apr 25 02:00:39.786 ERRO slog-async: logger dropped messages due to channel overflow, count: 1, producer_id: a6458b7d-87c3-4483-be96-854d814c20de, component: collection-task, collector_id: 39e6175b-4df2-4730-b11d-cbc1e60a2e78, component: oximeter-agent
```

From what I can tell, this means we are losing log messages.

If these log messages matter: we should consider mechanisms to batch them, [increase channel size](https://docs.rs/slog-async/latest/slog_async/struct.AsyncBuilder.html#method.chan_size), and avoid dropping them.

If these log messages don't matter: we should consider changing the log level or reducing the volume of overall messages.

Leaving the code as-is seems like the worst of both worlds - we don't really have a choice on "which" messages get dropped, so bumping into this limit regularly seems like it would limit the effectiveness of logging in oximeter overall.

**Comment by davepacheco** [comment]
Yeah, I think there might be a couple of problems here:

- We probably should be using synchronous logging everywhere?  I think we want to throttle activity rather than lose messages.
- If they should be async, yeah, we should increase the channel size.
- Why are we producing so many log messages?

**Comment by bnaecker** [comment]
I agree that this is not great. My recollection is that this is due to the [abnormally short metric collection interval](https://github.com/oxidecomputer/omicron/blob/db6bc39dac09a1c31464bd7874a075ddfbfee4e0/nexus/tests/integration_tests/oximeter.rs#L132) I used in some integration tests, just to keep them from taking forever. I have never seen this in a "real" situation. I'm not trying to say I shouldn't fix it, but I do think the problem at this point is the extraneous test output.

**Comment by jgallagher** [comment]
@bnaecker #1008 fixes this specific instance of dropped logs by effectively answering @davepacheco's third question, right? I.e., when the loop continued instead of returning, it would then start spamming debug logs, which is now fixed. Should we either leave this ticket open or open a separate one for his first two questions?

**Comment by bnaecker** [comment]
That's correct, thanks for clarifying @jgallagher. We're just spitting out a ton of debug messages that warn about the channel being closed, a condition which can never be recovered from, and so we'd get a tight-loop of `debug!` messages.

I think this particular issue was really about the spammy log output in tests, and should probably be closed. My sense is that the answer to the first of @davepacheco's other questions is "yes", we should avoid dropping messages whenever possible. I read up a bit on `slog-async`'s [overflow strategy](https://docs.rs/slog-async/latest/slog_async/enum.OverflowStrategy.html). We could conceivably use the `Block` variant to achieve that. I'd probably open a  new issue if we want to do that, but I'd defer to Dave on that.

The second point is probably not critical to open as a new issue. This particular flood of messages is fixed, and I've never seen this in another situation. It also seems like increasing the channel size is going to be more case-by-case, depending on how verbose the logging of each component is likely to be. Does that seem reasonable?

**Comment by jgallagher** [comment]
Yes, thanks. I was skimming the same strategies when it came up in chat earlier today, and agree switching to `Block` might be the best thing to try first (although without a spammy log source I'm not sure how we'd test/exercise that).

**Comment by bnaecker** [comment]
Ha, good point, I guess we can always revert this as a test case :)

**Comment by davepacheco** [comment]
I'm on board with trying `Block`!

**Comment by jgallagher** [comment]
I wrote up some notes for the general case in #1014; closing this as @bnaecker suggested as this specific error is now fixed.
